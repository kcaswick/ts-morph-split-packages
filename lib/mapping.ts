import * as fs from "fs-extra";
import { ICruiseResult } from "dependency-cruiser";

export type IConfigJson = {
  OldPatterns: Record<string, Location>;
};

type IMadgeDependencies = Record<string, string[]>;

function isICruiseResult(o: unknown): o is ICruiseResult {
  return typeof o === "object" && o !== null && "modules" in o && "summary" in o;
}

export type ILocation = {
  Repo: string;
  Package: string;
  Path: string;
  Order?: number;
};

class Location implements ILocation {
  Repo: string;
  Package: string;
  Path: string;
  Order?: number | undefined;
  /**
   *
   */
  constructor(source: ILocation) {
    this.Repo = source.Repo ?? "";
    this.Package = source.Package ?? "";
    this.Path = source.Path ?? "";
    this.Order = source.Order ?? 0;
  }

  public toString = () =>
    `${this.Package === this.Repo ? "" : `"${this.Package}"`}${this.Repo}:${this.Path}`;
}

export type IMapResult = {
  OldName: string;
  New?: Location;
};

export class MapResult implements IMapResult {
  OldName: string;
  New?: Location;

  /**
   *
   */
  constructor(source: IMapResult) {
    this.OldName = source.OldName;
    this.New = source.New;
  }

  public isUnmapped = () => this.New === undefined;

  public toString = () => `${this.OldName}${this.isUnmapped() ? " (unmapped)" : ` => ${this.New}`}`;
}

export class Config {
  OldPatterns = new Map<string, Location>();
  BranchPrefix = "split/";
}

// TODO: Switch to using this class for mappings
export class Dependency {
  Name: MapResult;
  dependencyMap: MapResult[] = [];
  /**
   *
   */
  constructor({
    Name,
    dependencyMap: dependencies,
  }: {
    Name: MapResult;
    dependencyMap: MapResult[];
  }) {
    this.Name = Name;
    this.dependencyMap = dependencies;
  }

  public toString = (): string =>
    `${this.Name} has ${
      this.dependencyMap.length > 0 ? this.dependencyMap.length : "no "
    }dependencies [${this.dependencyMap.map((d) => d.toString()).join(", ")}]`;
}

export class PackageMapping {
  config = new Config();
  depMap: {
    Name: MapResult;
    dependencyMap: MapResult[];
  }[] = [];

  invdepMap: {
    Name: MapResult;
    dependentMap: MapResult[];
  }[] = [];

  public export() {
    let output = this.depMap.map(($_) => ({
      OldName: $_.Name.OldName,
      NewRepo: $_.Name.New?.Repo,
      NewPackage: $_.Name.New?.Package,
      NewName: $_.Name.New?.Path,
      "Dependency Count": $_.dependencyMap.length,
      "Package Dependencies": $_.dependencyMap
        .map(($_) => $_.New?.Package ?? "N/A")
        .sort()
        .filter((value, index, self) => self.indexOf(value) === index)
        .filter((value) => value !== $_.Name.New?.Package),
      dependencies: $_.dependencyMap.map(($_) =>
        undefined === $_.New ? $_.OldName : $_.New?.Repo + ":" + $_.New?.Path
      ),
    }));
    output = output.sort((a, b) =>
      (`${a.NewRepo}${a.NewName}` ?? a.OldName ?? "").localeCompare(
        `${b.NewRepo}${b.NewName}` ?? b.OldName
      )
    );

    return JSON.stringify(output, undefined, 2);
  }

  public exportPackageDependenciesChart(): IMadgeDependencies {
    const output: IMadgeDependencies = {};
    this.depMap.forEach((x) => {
      const packageName =
        (x.Name.New?.Package === "N/A" ? x.Name.New?.Repo : x.Name.New?.Package) ?? "N/A";
      output[packageName] ??= [];
      x.dependencyMap
        .map((d) => (d.New?.Package === "N/A" ? d.New?.Repo : d.New?.Package) ?? "(unmapped)")
        .filter((d) => d !== packageName)
        .filter(
          (value, index, self) =>
            self.indexOf(value) === index &&
            !output[packageName].find((already) => value === already)
        )
        .forEach((d) => output[packageName].push(d));
    });

    return output;
  }

  public inverseDependencies(dependencyList: Record<string, string[]>) {
    const dependents: Record<string, string[]> = {};
    Object.entries(dependencyList).forEach((p) => {
      p[1].forEach((_: string | number) => {
        dependents[_] ??= [];
        dependents[_].push(p[0]);
      });
    });
    return dependents;
  }

  public mapPackage(oldPath: string): Location | undefined {
    for (const oldPattern of this.config.OldPatterns.keys()) {
      let target: ILocation | undefined;
      if (
        new RegExp(oldPattern).test(oldPath) &&
        (target = this.config.OldPatterns.get(oldPattern))
      ) {
        const nw: Location = new Location(target);
        nw.Path = oldPath.replace(oldPattern, nw.Path ?? "");
        return nw;
      }
    }

    return undefined;
  }

  public getPackageMap(
    dependencyJsonPath = "doc/dependency.json",
    configPath = "./PackageMap.json"
  ) {
    const packageMap: IConfigJson = fs.readJsonSync(configPath);

    const dependenciesJson: IMadgeDependencies | ICruiseResult =
      fs.readJsonSync(dependencyJsonPath);

    const dep2: Record<string, string[]> = isICruiseResult(dependenciesJson)
      ? Object.fromEntries(
          dependenciesJson.modules
            .filter(
              (s) =>
                s &&
                !(
                  s.source.startsWith("node_modules") ||
                  s.source.startsWith("@") ||
                  s.source.endsWith(".css")
                )
            )
            .map((m) => [
              m.source,
              m.dependencies
                .map((d) => d.resolved)
                .filter(
                  (s) =>
                    s && !(s.startsWith("node_modules") || s.startsWith("@") || s.endsWith(".css"))
                ) ?? [],
            ])
            .values()
        )
      : dependenciesJson;

    const dependents = this.inverseDependencies(dep2);

    Object.entries(packageMap.OldPatterns)
      .sort((a, b) => (a[1].Order ?? 0) - (b[1].Order ?? 0))
      .forEach((x) => this.config.OldPatterns.set(...x));

    this.depMap = Object.entries(dep2).map((x) => ({
      Name: new MapResult({ OldName: x[0], New: this.mapPackage(x[0]) }),
      dependencyMap: x[1].map((y) => new MapResult({ OldName: y, New: this.mapPackage(y) })),
    }));

    this.invdepMap = Object.entries(dependents).map((x) => ({
      Name: new MapResult({ OldName: x[0], New: this.mapPackage(x[0]) }),
      dependentMap: x[1].map((y) => new MapResult({ OldName: y, New: this.mapPackage(y) })),
    }));

    return this;
  }
}
