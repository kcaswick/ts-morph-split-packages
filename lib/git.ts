import { ensureDir } from "fs-extra";
import { basename, dirname, join } from "path";
import { BranchSummary, GitError, MoveResult, SimpleGit } from "simple-git";

import { PackageMapping } from "./mapping";

interface PromiseRejectedResult<E extends GitError = GitError> {
  status: "rejected";
  reason: E;
}

export const prepareGitMove = (
  mapping: PackageMapping,
  currentRepo: string = basename(process.cwd())
) => {
  const movesByRepo = new Map<string, Parameters<SimpleGit["mv"]>[]>();
  mapping.depMap
    .filter(
      (x) =>
        x.Name.isMapped() &&
        x.Name.New.Repo !== "N/A" &&
        x.Name.New.Repo !== currentRepo &&
        x.Name.OldName !== x.Name.New.Path
    )
    .forEach((x) => {
      if (x.Name.New) {
        let repoList = movesByRepo.get(x.Name.New.Repo);
        if (!repoList) {
          movesByRepo.set(x.Name.New.Repo, (repoList = []));
        }

        repoList.push([x.Name.OldName, x.Name.New.Path]);
      }
    });

  return movesByRepo;
};

const ensureFoldersExist = (targetPaths: string[]) => {
  console.debug(`${targetPaths.length} target paths: ${targetPaths.join(",")}`);
  const targetFolders = targetPaths
    .map((p) => dirname(p))
    .filter((value, index, self) => self.indexOf(value) === index)
    .sort((a, b) => (b.length === a.length ? a.localeCompare(b) : b.length - a.length));
  console.debug(`${targetFolders.length} target folders: ${targetFolders.join(",")}`);
  return Promise.all(targetFolders.map(ensureDir));
};

export const executeGitMoveForRepo = async (
  currentRepo: SimpleGit,
  targetRepo: string,
  moves: Parameters<SimpleGit["mv"]>[],
  mapping: PackageMapping
) => {
  await throwIfRepoNotReady(currentRepo);
  await currentRepo.checkoutLocalBranch(mapping.config.BranchPrefix + targetRepo);
  console.debug(`Removing files not in ${targetRepo}`);
  await removeFilesNotInTargetRepo(currentRepo, targetRepo, mapping);

  const throwMultipleFilesError = (x: Parameters<SimpleGit["mv"]>): string => {
    throw new Error(`Expected a single file but got list ${JSON.stringify(x[0])}`);
  };

  const ignored =
    moves.length > 0
      ? await currentRepo.checkIgnore(
          moves.map((x) => (typeof x[0] === "string" ? x[0] : throwMultipleFilesError(x)))
        )
      : [];
  console.debug(`Skipping ${ignored.length} git ignored files: ${ignored.join(",")}`);
  const movesInRepo = moves.filter(
    (x) => !ignored.includes(typeof x[0] === "string" ? x[0] : throwMultipleFilesError(x))
  );
  console.debug(`Moves remaining: ${movesInRepo.length}`);

  console.debug(`Ensuring destination folders exist for ${targetRepo}`);
  const currentRepoPath = await currentRepo.revparse("--absolute-git-dir");
  await ensureFoldersExist(movesInRepo.map((x) => join(currentRepoPath, "..", x[1])));
  console.debug(`Moving ${movesInRepo.length} files for ${targetRepo}`);
  const results = await Promise.allSettled(movesInRepo.map((x) => currentRepo.mv(...x)));
  const isPromiseFulfilled = <T>(p: PromiseSettledResult<T>) => p.status === "fulfilled";
  if (results.length > 0 && results.every(isPromiseFulfilled)) {
    try {
      await currentRepo.commit(`Move all files for ${targetRepo} to their new locations`);
    } catch (e) {
      const fulfilledPromises = results.filter(isPromiseFulfilled);
      console.debug(
        `${fulfilledPromises.length} fulfilled promises: ${JSON.stringify(fulfilledPromises)}`
      );
      throw e;
    }
  }

  // Throw if any failed.
  const rejectedPromises = results.filter(
    (p) => p.status === "rejected"
  ) as PromiseRejectedResult[];
  if (rejectedPromises.length > 0) {
    console.debug(
      `${rejectedPromises.length} rejected promises: ${JSON.stringify(rejectedPromises)}`
    );
    throw new AggregateError(rejectedPromises.map((p) => p.reason));
  }

  return results.map((p) => p.status === "fulfilled" && p.value);
};

export const executeGitMoveForRepos = async (
  currentRepo: SimpleGit,
  moves: Map<string, Parameters<SimpleGit["mv"]>[]>,
  mapping: PackageMapping,
  baseRepoName: string = basename(process.cwd())
) => {
  // Check that no branches are named split/
  console.log("Checking for split branches");
  let branches: BranchSummary;
  try {
    branches = await currentRepo.branch(["--list", "split/*"]);
  } catch (e) {
    console.error(e);
    throw e;
  }

  if (branches.all.length > 0) {
    throw new Error(
      `Current directory must not have any branches starting with split/, but found branches ${branches.all.join(
        ","
      )}`
    );
  }

  const startCommitish = await currentRepo.revparse(["HEAD"]);

  let moveResults = await Array.from(moves.entries()).reduce(
    async (lastProm, [repoPath, repoMoves]) => {
      const resultSoFar: (false | MoveResult)[] = await lastProm;
      const results: (false | MoveResult)[] = await executeGitMoveForRepo(
        currentRepo,
        repoPath,
        repoMoves,
        mapping
      );
      await currentRepo.checkout(startCommitish);
      return [...resultSoFar, ...results];
    },
    Promise.resolve<(false | MoveResult)[]>([])
  );

  const baseMoves: Parameters<SimpleGit["mv"]>[] = mapping.depMap
    .filter(
      (x) =>
        x.Name.isMapped() &&
        (x.Name.New.Repo === "N/A" || x.Name.New.Repo === baseRepoName) &&
        x.Name.OldName !== x.Name.New.Path
    )
    .map((x) => [x.Name.OldName, x.Name.New?.Path ?? ""]);
  const baseMoveResults = await executeGitMoveForRepo(
    currentRepo,
    baseRepoName,
    baseMoves,
    mapping
  );
  moveResults = moveResults.concat(baseMoveResults);

  return moveResults;
};

/**
 * Removes all files that are not part of the target repository.
 * @param currentRepo The git repo that we are currently on.
 * @param targetRepo The name of the repo that we are splitting to.
 * @param mapping The list of packages that are in the repository.
 */
async function removeFilesNotInTargetRepo(
  currentRepo: SimpleGit,
  targetRepo: string,
  mapping: PackageMapping
) {
  const filesToRemove = mapping.depMap
    .filter((x) => x.Name.isMapped() && x.Name.New.Repo !== targetRepo)
    .map((x) => x.Name.OldName);
  const ignored = await currentRepo.checkIgnore(filesToRemove);
  const gitFilesToRemove = filesToRemove.filter((x) => !ignored.includes(x));
  console.debug(`${gitFilesToRemove.length} files to remove: ${gitFilesToRemove.join(",")}`);
  await currentRepo.rm(gitFilesToRemove);
  await currentRepo.commit(`Remove all files that are not part of ${targetRepo}`);
}

const throwIfRepoNotReady = async (currentRepo: SimpleGit) => {
  if (!(await currentRepo.checkIsRepo())) {
    throw new Error(`Current directory must be a git working tree`);
  }

  // Check that no uncommitted changes, etc.
  const status = await currentRepo.status();
  if (!status.isClean()) {
    throw new Error(
      `Current directory must not have any uncommitted changes, but found ${
        status.files.length
      }: ${status.files.map((f) => f.path)}`
    );
  }
};
