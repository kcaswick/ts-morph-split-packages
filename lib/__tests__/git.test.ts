/**
 *
 * Tests for lib\git.ts
 *
 */
import { existsSync } from "fs";
import { join } from "path";
import simpleGit, { CheckRepoActions } from "simple-git";
import temp from "temp";

import * as sut from "../git";
import { MapResult, PackageMapping } from "../mapping";
import {
  checkoutTempSimpleRepo,
  createTemporaryRepository,
  loadSimpleMadge,
} from "./test_fixtures";
describe("prepareGitMove", () => {
  it("Expect to not log errors in console", () => {
    const spy = jest.spyOn(global.console, "error");
    const m = loadSimpleMadge();
    sut.prepareGitMove(m);
    expect(spy).not.toHaveBeenCalled();
  });
  it("Should run and match the snapshot", () => {
    const m = loadSimpleMadge();
    const result = sut.prepareGitMove(m);
    expect(result).toMatchSnapshot();
  });
  it("should move some files to new, and leave others", () => {
    const m = loadSimpleMadge();
    const result = sut.prepareGitMove(m);
    const newRepo = result.get("new");
    expect(newRepo).toBeDefined();
    expect(newRepo).toContainEqual(["lib/mapping.ts", "src/mapping.ts"]);
    expect(newRepo).not.toContainEqual(["lib/index.ts", expect.any(String)]);
  });
  it("No moves needed", () => {
    const mapping = Object.assign(new PackageMapping(), {
      depMap: [
        {
          Name: new MapResult({
            OldName: "ts-morph-split-packages",
            New: {
              Repo: "N/A",
              Package: "N/A",
              Path: "ts-morph-split-packages",
            },
          }),
        },
      ],
    } as PackageMapping);
    const currentRepo = "ts-morph-split-packages";
    const movesByRepo = sut.prepareGitMove(mapping, currentRepo);
    expect(movesByRepo.size).toStrictEqual(0);
  });
});
describe("executeGitMoveForRepo", () => {
  const repoPromise = checkoutTempSimpleRepo();

  const spy = jest.spyOn(global.console, "error");
  const m = loadSimpleMadge();
  const plan = sut.prepareGitMove(m); // For a snapshot, see `prepareGitMove Should run and match the snapshot 1`
  expect(plan.get("new")).toBeDefined();
  expect(spy).not.toHaveBeenCalled();

  it("checkoutTempSimpleRepo", async () => {
    const [tempRepoPath, tempRepo] = await repoPromise;
    expect(existsSync(tempRepoPath));
    await expect(tempRepo.checkIsRepo(CheckRepoActions.IS_REPO_ROOT)).resolves.toBeTruthy();
  });

  it("Execute default plan", async () => {
    const [tempRepoPath, tempRepo] = await repoPromise;
    const startCommitish = await tempRepo.revparse(["HEAD"]);
    const results = expect(
      sut.executeGitMoveForRepo(tempRepo, "new", plan.get("new")!, m)
    ).resolves;
    await results.toMatchSnapshot();
    expect(existsSync(join(tempRepoPath, "lib/index.ts"))).toBeFalsy();
    expect(existsSync(join(tempRepoPath, "src/mapping.ts"))).toBeTruthy();

    await tempRepo.checkout(startCommitish);

    const resultsTestFixtures = expect(
      sut.executeGitMoveForRepo(tempRepo, "test_fixtures", plan.get("test_fixtures")!, m)
    ).resolves;
    await resultsTestFixtures.toMatchSnapshot();
    expect(existsSync(join(tempRepoPath, "src/__tests__/test_fixtures.ts"))).toBeTruthy();
  }, 15000);
});
describe("executeGitMoveForRepos", () => {
  it("Should throw if branch split/any is present", async () => {
    const [_tempRepoPath, tempRepo] = await createTemporaryRepository("empty_folder");
    await expect(tempRepo.commit("Empty initial commit", ["--allow-empty"])).resolves.toBeDefined();
    await expect(tempRepo.branch(["split/any"])).resolves.toBeDefined();
    const m = new PackageMapping(); // X loadSimpleMadge();
    await expect(sut.executeGitMoveForRepos(tempRepo, new Map(), m)).rejects.toThrow(/split/i);
    /* .toThrowErrorMatchingInlineSnapshot() */
  });
  it("Complete simple split", async () => {
    const [tempRepoPath, tempRepo] = await checkoutTempSimpleRepo();
    const m = loadSimpleMadge();
    const plan = sut.prepareGitMove(m);
    expect(plan).toMatchSnapshot();
    const results = expect(sut.executeGitMoveForRepos(tempRepo, plan, m)).resolves;

    await results.toMatchSnapshot();

    expect(existsSync(join(tempRepoPath, "lib/index.ts"))).toBeTruthy();
    expect(existsSync(join(tempRepoPath, "lib/mapping.ts"))).toBeFalsy();
    expect(existsSync(join(tempRepoPath, "lib/__tests__/test_fixtures.ts"))).toBeFalsy();
  }, 15000);
});
describe("throwIfRepoNotReady", () => {
  it("Should throw if dir is not a repo", async () => {
    const tempRepoPath = await temp.mkdir("empty_folder");
    const tempRepo = simpleGit(tempRepoPath);
    const m = new PackageMapping(); // X loadSimpleMadge();
    await expect(sut.executeGitMoveForRepo(tempRepo, "fake", [], m)).rejects.toThrow(
      /fatal: (not a git repository (or any of the parent directories): |invalid gitfile format:.*)\.git/i
    );
  }, 15000);
});
