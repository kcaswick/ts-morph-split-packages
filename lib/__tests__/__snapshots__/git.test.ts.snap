// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`executeGitMoveForRepo Execute default plan 1`] = `
Map {
  "test_fixtures" => Array [
    Array [
      "lib/__tests__/test_fixtures.ts",
      "src/__tests__/test_fixtures.ts",
    ],
  ],
  "new" => Array [
    Array [
      "lib/mapping.ts",
      "src/mapping.ts",
    ],
  ],
}
`;

exports[`executeGitMoveForRepo Execute default plan 2`] = `
Array [
  Object {
    "moves": Array [
      Object {
        "from": "lib/mapping.ts",
        "to": "src/mapping.ts",
      },
    ],
  },
]
`;

exports[`executeGitMoveForRepo Execute default plan 3`] = `
Array [
  Object {
    "moves": Array [
      Object {
        "from": "lib/__tests__/test_fixtures.ts",
        "to": "src/__tests__/test_fixtures.ts",
      },
    ],
  },
]
`;

exports[`executeGitMoveForRepos Complete simple split 1`] = `
Map {
  "test_fixtures" => Array [
    Array [
      "lib/__tests__/test_fixtures.ts",
      "src/__tests__/test_fixtures.ts",
    ],
  ],
  "new" => Array [
    Array [
      "lib/mapping.ts",
      "src/mapping.ts",
    ],
  ],
}
`;

exports[`executeGitMoveForRepos Complete simple split 2`] = `
Array [
  Object {
    "moves": Array [
      Object {
        "from": "lib/__tests__/test_fixtures.ts",
        "to": "src/__tests__/test_fixtures.ts",
      },
    ],
  },
  Object {
    "moves": Array [
      Object {
        "from": "lib/mapping.ts",
        "to": "src/mapping.ts",
      },
    ],
  },
]
`;

exports[`prepareGitMove Should run and match the snapshot 1`] = `
Map {
  "test_fixtures" => Array [
    Array [
      "lib/__tests__/test_fixtures.ts",
      "src/__tests__/test_fixtures.ts",
    ],
  ],
  "new" => Array [
    Array [
      "lib/mapping.ts",
      "src/mapping.ts",
    ],
  ],
}
`;
